p2p/router.go

```diff
@@ -0,0 +1,557 @@
+// the router is shutting down.
+func (r *Router) routeChannel(channel *Channel) {
+	for {
+		select {
+		case envelope, ok := <-channel.outCh:
+			if !ok {
+				return
+			}
+
+			// FIXME: This is a bit unergonomic, maybe it'd be better for Wrap()
+			// to return a wrapped copy.
+			if _, ok := channel.messageType.(Wrapper); ok {
+				wrapper := proto.Clone(channel.messageType)
+				if err := wrapper.(Wrapper).Wrap(envelope.Message); err != nil {
+					r.Logger.Error("failed to wrap message", "err", err)
+					continue
+				}
+				envelope.Message = wrapper
+			}
+			envelope.channel = channel.id
+
+			if envelope.Broadcast {
+				r.peerMtx.RLock()
+				peerQueues := make(map[NodeID]queue, len(r.peerQueues))
+				for peerID, peerQueue := range r.peerQueues {
+					peerQueues[peerID] = peerQueue
+				}
+				r.peerMtx.RUnlock()
+
+				for peerID, peerQueue := range peerQueues {
+					e := envelope
+					e.Broadcast = false
+					e.To = peerID
+					select {
+					case peerQueue.enqueue() <- e:
+					case <-peerQueue.closed():
+					case <-r.stopCh:
+						return
+					}
+				}
+
+			} else {
+				r.peerMtx.RLock()
+				peerQueue, ok := r.peerQueues[envelope.To]
+				r.peerMtx.RUnlock()
+				if !ok {
+					r.logger.Error("dropping message for non-connected peer", "peer", envelope.To)
+					continue
+				}
+
+				select {
+				case peerQueue.enqueue() <- envelope:
+				case <-peerQueue.closed():
+					r.logger.Error("dropping message for non-connected peer", "peer", envelope.To)
+				case <-r.stopCh:
+					return
+				}
+			}
+
+		case peerError, ok := <-channel.errCh:
+			if !ok {
+				return
+			}
+			// FIXME: We just disconnect the peer for now
+			r.logger.Error("peer error, disconnecting", "peer", peerError.PeerID, "err", peerError.Err)
+			r.peerMtx.RLock()
+			peerQueue, ok := r.peerQueues[peerError.PeerID]
+			r.peerMtx.RUnlock()
+			if ok {
+				peerQueue.close()
+			}
+
+		case <-channel.Done():
+			return
+		case <-r.stopCh:
+			return
+		}
+	}
+}
+
+// acceptPeers accepts inbound connections from peers on the given transport.
+func (r *Router) acceptPeers(transport Transport) {
+	for {
+		select {
+		case <-r.stopCh:
+			return
+		default:
+		}
+
+		conn, err := transport.Accept(context.Background())
+		switch err {
+		case nil:
+		case ErrTransportClosed{}, io.EOF:
+			r.logger.Info("transport closed, stopping accept routine", "transport", transport)
+			return
+		default:
+			r.logger.Error("failed to accept connection", "transport", transport, "err", err)
+			return
+		}
+
+		peerID := conn.NodeInfo().NodeID
+		if r.store.Claim(peerID) == nil {
+			r.logger.Error("already connected to peer, rejecting connection", "peer", peerID)
+			_ = conn.Close()
+			continue
+		}
+
+		queue := newFIFOQueue()
+		r.peerMtx.Lock()
+		r.peerQueues[peerID] = queue
+		r.peerMtx.Unlock()
+
+		go func() {
+			defer func() {
+				r.peerMtx.Lock()
+				delete(r.peerQueues, peerID)
+				r.peerMtx.Unlock()
+				queue.close()
```

Comment:
Don't we also close this guy in `routeChannel`?

Reply:
Yes, when we receive peer errors -- that's fine, `close()` is idempotent. We have to close it when the peer goroutine exits, but we also use queue closure as a signal from other parts of the router to disconnect a peer. The queue is the only shared peer state in the router.